ruby_02-web_applications_with_ruby/outlines/intro_to_crud_redux.markdown

```diff
@@ -0,0 +1,174 @@
+* divide responsibilities between the controller, views, and models
+
+### Structure
+
+* Lecture
+* Code-along
+* Work Time
+
+### Repository
+
+* If you have been following the [TaskManager](https://github.com/s-espinosa/task_manager_redux) tutorials up to this point, you should be able to use the version you already have.
+* If not, clone the [TaskManager Repo](https://github.com/s-espinosa/task_manager_redux), and switch to the crud-lesson branch.
+
+### Warm Up
+
+* What does CRUD stand for and why is it important?
+* What does MVC stand for? Create a diagram of the MVC pattern.
+* What part of the MVC does `task_manager_app.rb` represent in our current TaskManager app? What about `task.rb`?
+
+### CRUD Introduction
+
+Assuming we want to create full CRUD functionality in our Sinatra app for users accessing our site through a browser, there are seven routes that we will need to define. Let's make a chart together.
+
+(Here's a [completed chart](https://www.dropbox.com/s/vx3ocfsusjdrgfw/crud_in_sinatra.pdf?dl=0), but don't open it until you're finished!)
+
+### Code-Along
+
+We're going to follow the MVC design pattern (Rails uses this by default, but in Sinatra we will need to create this structure ourselves).
+
+So far, we are able to create a task and read a task. How do we add functionality to update and delete tasks?
+
+#### Editing a task
+
+Our **edit** route should bring the user to a form where they can change the title and description of the task. In order to create this functionality, we'll need to create a button to get to the edit page, a route in our controller for that link, and a view that will be rendered when we hit the new route.
+
+In our `index.erb` view:
+
+```erb
+<h1>All Tasks</h1>
+
+<% @tasks.each do |task| %>
+  <h3><a href="/tasks/<%= task.id %>"><%= task.title %></a></h3>
+  <p><%= task.description  %></p>
+  <a href="/tasks/<%= task.id %>/edit">Edit</a>
+<% end %>
+```
+
+In our controller:
+
+```ruby
+  get '/tasks/:id/edit' do
+    @task = Task.find(params[:id])
+    erb :edit
+  end
+```
+In our view, `edit.erb`:
+
+```erb
+<form action="/tasks/<%= @task.id %>" method="post">
+  <p>Edit</p>
+  <input type="hidden" name="_method" value="PUT" />
+  <input type='text' name='task[title]' value="<%= @task.title %>"/><br/>
+  <textarea name='task[description]'><%= @task.description %></textarea><br/>
+  <input type='submit'/>
+</form>
+```
+
+#### Updating a task
+
+Our new form needs somewhere to go when a user clicks submit. We'll use the **update** route to do the work of changing the task in the database.
+
+In our controller:
+
+```ruby
+  set :method_override, true  # this allows us to use _method in the form
+  ...
+  put '/tasks/:id' do |id|
+    Task.update(id.to_i, params[:task])
+    redirect "/tasks/#{id}"
+  end
+```
+
+In our TaskManager model:
+
+```ruby
+  def self.update(id, task_params)
+    database = SQLite3::Database.new('db/task_manager_development.db')
+    database.results_as_hash = true
```

Comment:
This, along with line 99 have been taken care of in `#initialize`, right? So we could just reference `@database` on line 101?